File: C++/Valid Parentheses.cpp
#include <vector>

class Solution {
public:
    bool isValid(string s) {

        std::vector<char> stack;
        
        if (s.length() % 2 != 0) return false;

        for (auto x : s) {

            if (x == ')') {
                if (stack.size() == 0 || stack.back() != '(') {
                    return false;
                }
                stack.pop_back();
            } else if (x == ']') {
                if (stack.size() == 0 || stack.back() != '[') {
                    return false;
                }
                stack.pop_back();
            } else if (x == '}') {
                if (stack.size() == 0 || stack.back() != '{') {
                    return false;
                }
                stack.pop_back();
            }
            
            if (x == '(' || x == '[' || x == '{')
                stack.push_back(x);

        }

        return !(stack.size() > 0);
    }
};

File: python/Valid Parentheses.py
class Solution:

    def isValid(self, s: str) -> bool:
        
        stack = []

        closing = 0
        opening = 0

        if len(s) % 2 != 0:
            return False

        for x in s:
            if x == ')' and (len(stack) == 0 or stack.pop() != '('):
                return False
            elif x == ']' and (len(stack) == 0 or stack.pop() != '['):
                return False
            if x == '}' and (len(stack) == 0 or stack.pop() != '{'):
                return False
            
            if x == '(' or x == '[' or x == '{':
                stack.append(x)

        if len(stack) > 0:
            return False

        return True
